### PR TITLE
Add Cholesky, QR, and Triangular solve implementations.

### PR DESCRIPTION
**Pull request by hawkinsp**
* Adds lax.{cholesky,triangular_solve,qr}. Adds a JVP for Cholesky.
* Adds a transpose rule for add_p, needed by the Cholesky JVP.
* Adds np.linalg.{cholesky,qr,dot,matmul,trace}.
* Adds scipy.linalg.{cholesky,qr,solve_triangular,tril,triu}.

Pair programmed with mattjj.

Partially addresses #44

### CODE REVIEW

**Comment by hawkinsp** [comment]
This doesn't add tests for the JVPs yet. I wanted to get this in as is before working on tests for the gradients.

File: jax/scipy/linalg.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+  return q, r
+
+
+@_wraps(scipy.linalg.solve_triangular)
+def solve_triangular(a, b, trans=0, lower=False, unit_diagonal=False,
+                     overwrite_b=False, debug=None, check_finite=True):
+  del overwrite_b, debug, check_finite
+
+  if unit_diagonal:
+    raise NotImplementedError("unit_diagonal=True is not implemented.")
```
Comment (mattjj) [inline jax/scipy/linalg.py]: IIUC exploiting the unit diagonal is just an optimization; it should still work if we ignore the setting of this argument, right?

File: jax/scipy/linalg.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  q, r = lax_linalg.qr(a, full_matrices)
+  if mode == "r":
+    return r
+  return q, r
+
+
+@_wraps(scipy.linalg.solve_triangular)
+def solve_triangular(a, b, trans=0, lower=False, unit_diagonal=False,
+                     overwrite_b=False, debug=None, check_finite=True):
+  del overwrite_b, debug, check_finite
```
Comment (mattjj) [inline jax/scipy/linalg.py]: We might want to warn if these flags are set to anything but their default values (by which I mean, set in any way that doesn't match the code that follows).

**Review by mattjj — APPROVED** [review]
Awesome! (For the record, I only really paired for the cholesky and triangular_solve JVPs/transposes, so most of this is @hawkinsp's work!)

We could implement a general inv on top of QR. I know LU is often preferred, but on some backends it might even be faster to use QR. Just a thought.

These will require a new jaxlib release, right?

Minor comments on a couple lines. Feel free to ignore.

**Pull request merged**